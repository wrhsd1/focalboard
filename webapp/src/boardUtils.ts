// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Utils} from './utils'
import {Card} from './blocks/card'
import {IPropertyTemplate, IPropertyOption, BoardGroup} from './blocks/board'

export function groupCardsByOptions(cards: Card[], optionIds: string[], groupByProperty?: IPropertyTemplate): BoardGroup[] {
    const groups = []
    for (const optionId of optionIds) {
        if (optionId) {
            const option = groupByProperty?.options.find((o) => o.id === optionId)
            if (option) {
                const c = cards.filter((o) => optionId === o.fields?.properties[groupByProperty!.id])
                const group: BoardGroup = {
                    option,
                    cards: c,
                }
                groups.push(group)
            } else {
                Utils.logError(`groupCardsByOptions: Missing option with id: ${optionId}`)
            }
        } else {
            // Empty group
            const emptyGroupCards = cards.filter((card) => {
                const groupByOptionId = card.fields.properties[groupByProperty?.id || '']
                return !groupByOptionId || !groupByProperty?.options.find((option) => option.id === groupByOptionId)
            })
            const group: BoardGroup = {
                option: {id: '', value: `No ${groupByProperty?.name}`, color: ''},
                cards: emptyGroupCards,
            }
            groups.push(group)
        }
    }
    return groups
}

export function getVisibleAndHiddenGroups(cards: Card[], visibleOptionIds: string[], hiddenOptionIds: string[], groupByProperty?: IPropertyTemplate): {visible: BoardGroup[], hidden: BoardGroup[]} {
    let unassignedOptionIds: string[] = []
    if (groupByProperty) {
        unassignedOptionIds = groupByProperty.options.
            filter((o: IPropertyOption) => !visibleOptionIds.includes(o.id) && !hiddenOptionIds.includes(o.id)).
            map((o: IPropertyOption) => o.id)
    }
    const allVisibleOptionIds = [...visibleOptionIds, ...unassignedOptionIds]

    // If the empty group positon is not explicitly specified, make it the first visible column
    if (!allVisibleOptionIds.includes('') && !hiddenOptionIds.includes('')) {
        allVisibleOptionIds.unshift('')
    }

    const visibleGroups = groupCardsByOptions(cards, allVisibleOptionIds, groupByProperty)
    const hiddenGroups = groupCardsByOptions(cards, hiddenOptionIds, groupByProperty)
    return {visible: visibleGroups, hidden: hiddenGroups}
}
