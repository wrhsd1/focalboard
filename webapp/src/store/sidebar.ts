// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {createAsyncThunk, createSelector, createSlice, PayloadAction} from '@reduxjs/toolkit'

import {default as client} from '../octoClient'

import {RootState} from './index'

interface Category {
    id: string
    name: string
    userID: string
    teamID: string
    createAt: number
    updateAt: number
    deleteAt: number
    collapsed: boolean
}

interface CategoryBoards extends Category {
    boardIDs: Array<string>
}

interface BoardCategoryWebsocketData {
    boardID: string
    categoryID: string
}

export const DefaultCategory: CategoryBoards = {
    id: '',
    name: 'Boards',
} as CategoryBoards

export const fetchSidebarCategories = createAsyncThunk(
    'sidebarCategories/fetch',
    async (teamID: string) => {
        const categories = await client.getSidebarCategories(teamID)
        return categories.sort((a, b) => a.name.localeCompare(b.name))
    },
)

type Sidebar = {
    categoryAttributes: Array<CategoryBoards>
}

const sidebarSlice = createSlice({
    name: 'sidebar',
    initialState: {categoryAttributes: []} as Sidebar,
    reducers: {
        updateCategories: (state, action: PayloadAction<Array<Category>>) => {
            action.payload.forEach((updatedCategory) => {
                const index = state.categoryAttributes.findIndex((c) => c.id === updatedCategory.id)

                // when new category got created
                if (index === -1) {
                    state.categoryAttributes.push({
                        ...updatedCategory,
                        boardIDs: [],
                    })
                } else if (updatedCategory.deleteAt) {
                    // when category is deleted
                    state.categoryAttributes.splice(index, 1)
                } else {
                    // else all, update the category
                    state.categoryAttributes[index] = {
                        ...state.categoryAttributes[index],
                        name: updatedCategory.name,
                        updateAt: updatedCategory.updateAt,
                    }
                }
            })

            // sort categories alphabetically
            state.categoryAttributes.sort((a, b) => a.name.localeCompare(b.name))
        },
        updateBoardCategories: (state, action: PayloadAction<Array<BoardCategoryWebsocketData>>) => {
            action.payload.forEach((boardCategory) => {
                for (let i = 0; i < state.categoryAttributes.length; i++) {
                    const categoryAttribute = state.categoryAttributes[i]

                    // first we remove the board from list of boards
                    categoryAttribute.boardIDs = categoryAttribute.boardIDs.filter((boardID) => boardID !== boardCategory.boardID)

                    // then we add it if this is the target category
                    if (categoryAttribute.id === boardCategory.categoryID) {
                        categoryAttribute.boardIDs.push(boardCategory.boardID)
                    }
                }
            })
        },
    },
    extraReducers: (builder) => {
        builder.addCase(fetchSidebarCategories.fulfilled, (state, action) => {
            state.categoryAttributes = action.payload || []
        })
    },
})

export const getSidebarCategories = createSelector(
    (state: RootState): Array<CategoryBoards> => state.sidebar.categoryAttributes,
    (sidebarCategories) => sidebarCategories,
)

export const {reducer} = sidebarSlice

export const {updateCategories, updateBoardCategories} = sidebarSlice.actions

export {Category, CategoryBoards, BoardCategoryWebsocketData}

