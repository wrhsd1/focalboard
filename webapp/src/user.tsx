// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

interface IUser {
    id: string,
    username: string,
    email: string,
    nickname: string,
    firstname: string,
    lastname: string,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    props: Record<string, any>,
    create_at: number,
    update_at: number,
    is_bot: boolean,
    roles: string,
}

interface UserWorkspace {
    id: string
    title: string
    boardCount: number
}

interface UserConfigPatch {
    updatedFields?: Record<string, string>
    deletedFields?: string[]
}

function parseUserProps(props: Record<string, any>): Record<string, any> {
    const processedProps = props
    const hiddenBoardIDs = props.hiddenBoardIDs ? JSON.parse(props.hiddenBoardIDs) : []
    processedProps.hiddenBoardIDs = {}
    hiddenBoardIDs.forEach((boardID: string) => processedProps.hiddenBoardIDs[boardID] = true)
    return processedProps
}

const UserPropPrefix = 'focalboard_'

export {
    IUser,
    UserWorkspace,
    UserConfigPatch,
    UserPropPrefix,
    parseUserProps,
}
