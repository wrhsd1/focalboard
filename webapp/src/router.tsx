// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React, {useEffect, useMemo} from 'react'
import {
    Router,
    Redirect,
    Switch,
    useRouteMatch,
    useHistory,
    generatePath,
    useLocation,
} from 'react-router-dom'
import {createBrowserHistory, History} from 'history'

import {IAppWindow} from './types'
import BoardPage from './pages/boardPage/boardPage'
import ChangePasswordPage from './pages/changePasswordPage'
import WelcomePage from './pages/welcome/welcomePage'
import ErrorPage from './pages/errorPage'
import LoginPage from './pages/loginPage'
import RegisterPage from './pages/registerPage'
import {Utils} from './utils'
import octoClient from './octoClient'
import {setGlobalError, getGlobalError} from './store/globalError'
import {useAppSelector, useAppDispatch} from './store/hooks'
import {getFirstTeam, fetchTeams, Team} from './store/teams'
import {UserSettings} from './userSettings'
import FBRoute from './route'

declare let window: IAppWindow

const UUID_REGEX = new RegExp(/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/)

function HomeToCurrentTeam(props: {path: string, exact: boolean}) {
    return (
        <FBRoute
            path={props.path}
            exact={props.exact}
            loginRequired={true}
            component={() => {
                const firstTeam = useAppSelector<Team|null>(getFirstTeam)
                const dispatch = useAppDispatch()
                useEffect(() => {
                    dispatch(fetchTeams())
                }, [])

                let teamID = (window.getCurrentTeamId && window.getCurrentTeamId()) || ''
                const lastTeamID = UserSettings.lastTeamId
                if (!teamID && !firstTeam && !lastTeamID) {
                    return <></>
                }
                teamID = teamID || lastTeamID || firstTeam?.id || ''

                if (UserSettings.lastBoardId) {
                    const lastBoardID = UserSettings.lastBoardId[teamID]
                    const lastViewID = UserSettings.lastViewId[lastBoardID]

                    if (lastBoardID && lastViewID) {
                        return <Redirect to={`/team/${teamID}/${lastBoardID}/${lastViewID}`}/>
                    }
                    if (lastBoardID) {
                        return <Redirect to={`/team/${teamID}/${lastBoardID}`}/>
                    }
                }

                return <Redirect to={`/team/${teamID}`}/>
            }}
        />
    )
}

function WorkspaceToTeamRedirect() {
    const match = useRouteMatch<{boardId: string, viewId: string, cardId?: string, workspaceId?: string}>()
    const queryParams = new URLSearchParams(useLocation().search)
    const history = useHistory()
    useEffect(() => {
        octoClient.getBoard(match.params.boardId).then((board) => {
            if (board) {
                let newPath = generatePath(match.path.replace('/workspace/:workspaceId', '/team/:teamId'), {
                    teamId: board?.teamId,
                    boardId: board?.id,
                    viewId: match.params.viewId,
                    cardId: match.params.cardId,
                })
                if (queryParams) {
                    newPath += '?' + queryParams
                }
                history.replace(newPath)
            }
        })
    }, [])
    return null
}

function GlobalErrorRedirect() {
    const globalError = useAppSelector<string>(getGlobalError)
    const dispatch = useAppDispatch()
    const history = useHistory()

    useEffect(() => {
        if (globalError) {
            dispatch(setGlobalError(''))
            history.replace(`/error?id=${globalError}`)
        }
    }, [globalError, history])

    return null
}

type Props = {
    history?: History<unknown>
}

const FocalboardRouter = (props: Props): JSX.Element => {
    const isPlugin = Utils.isFocalboardPlugin()

    let browserHistory: History<unknown>
    if (props.history) {
        browserHistory = props.history
    } else {
        browserHistory = useMemo(() => {
            return createBrowserHistory({basename: Utils.getFrontendBaseURL()})
        }, [])
    }

    if (isPlugin) {
        useEffect(() => {
            if (window.frontendBaseURL) {
                browserHistory.replace(window.location.pathname.replace(window.frontendBaseURL, ''))
            }
        }, [])
    }

    return (
        <Router history={browserHistory}>
            <GlobalErrorRedirect/>
            <Switch>
                {isPlugin &&
                    <HomeToCurrentTeam
                        path='/'
                        exact={true}
                    />}
                {isPlugin &&
                    <FBRoute
                        exact={true}
                        path='/welcome'
                    >
                        <WelcomePage/>
                    </FBRoute>}

                <FBRoute path='/error'>
                    <ErrorPage/>
                </FBRoute>

                {!isPlugin &&
                    <FBRoute path='/login'>
                        <LoginPage/>
                    </FBRoute>}
                {!isPlugin &&
                    <FBRoute path='/register'>
                        <RegisterPage/>
                    </FBRoute>}
                {!isPlugin &&
                    <FBRoute path='/change_password'>
                        <ChangePasswordPage/>
                    </FBRoute>}

                <FBRoute path={['/team/:teamId/new/:channelId']}>
                    <BoardPage new={true}/>
                </FBRoute>

                <FBRoute path={['/team/:teamId/shared/:boardId?/:viewId?/:cardId?', '/shared/:boardId?/:viewId?/:cardId?']}>
                    <BoardPage readonly={true}/>
                </FBRoute>

                <FBRoute
                    loginRequired={true}
                    path='/board/:boardId?/:viewId?/:cardId?'
                    getOriginalPath={({params: {boardId, viewId, cardId}}) => {
                        return `/board/${Utils.buildOriginalPath('', boardId, viewId, cardId)}`
                    }}
                >
                    <BoardPage/>
                </FBRoute>
                <FBRoute path={['/workspace/:workspaceId/shared/:boardId?/:viewId?/:cardId?', '/workspace/:workspaceId/:boardId?/:viewId?/:cardId?']}>
                    <WorkspaceToTeamRedirect/>
                </FBRoute>
                <FBRoute
                    loginRequired={true}
                    path='/team/:teamId/:boardId?/:viewId?/:cardId?'
                    getOriginalPath={({params: {teamId, boardId, viewId, cardId}}) => {
                        return `/team/${Utils.buildOriginalPath(teamId, boardId, viewId, cardId)}`
                    }}
                >
                    <BoardPage/>
                </FBRoute>

                {!isPlugin &&
                    <FBRoute
                        path='/:boardId?/:viewId?/:cardId?'
                        loginRequired={true}
                        getOriginalPath={({params: {boardId, viewId, cardId}}) => {
                            const boardIdIsValidUUIDV4 = UUID_REGEX.test(boardId || '')
                            if (boardIdIsValidUUIDV4) {
                                return `/${Utils.buildOriginalPath('', boardId, viewId, cardId)}`
                            }
                            return ''
                        }}
                    >
                        <BoardPage/>
                    </FBRoute>}
            </Switch>
        </Router>
    )
}

export default React.memo(FocalboardRouter)
