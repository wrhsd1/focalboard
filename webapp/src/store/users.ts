// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {createSlice, createAsyncThunk, PayloadAction, createSelector} from '@reduxjs/toolkit'

import {default as client} from '../octoClient'
import {IUser, parseUserProps} from '../user'

import {Utils} from '../utils'

import {Subscription} from '../wsclient'

// TODO: change this whene the initial load is complete
// import {initialLoad} from './initialLoad'
import {UserSettings} from '../userSettings'

import {RootState} from './index'

export const fetchMe = createAsyncThunk(
    'users/fetchMe',
    async () => client.getMe(),
)

export const versionProperty = 'focalboard_version72MessageCanceled'

type UsersStatus = {
    me: IUser|null
    boardUsers: {[key: string]: IUser}
    loggedIn: boolean|null
    blockSubscriptions: Array<Subscription>
}

export const fetchUserBlockSubscriptions = createAsyncThunk(
    'user/blockSubscriptions',
    async (userId: string) => (Utils.isFocalboardPlugin() ? client.getUserBlockSubscriptions(userId) : []),
)

const initialState = {
    me: null,
    boardUsers: {},
    loggedIn: null,
    userWorkspaces: [],
    blockSubscriptions: [],
} as UsersStatus

const usersSlice = createSlice({
    name: 'users',
    initialState,
    reducers: {
        setMe: (state, action: PayloadAction<IUser|null>) => {
            state.me = action.payload
            if (state.me) {
                state.me.props = parseUserProps(state.me.props)
            }
            state.loggedIn = Boolean(state.me)
        },
        setBoardUsers: (state, action: PayloadAction<IUser[]>) => {
            state.boardUsers = action.payload.reduce((acc: {[key: string]: IUser}, user: IUser) => {
                acc[user.id] = user
                return acc
            }, {})
        },
        addBoardUsers: (state, action: PayloadAction<IUser[]>) => {
            action.payload.forEach((user: IUser) => {
                state.boardUsers[user.id] = user
            })
        },
        removeBoardUsersById: (state, action: PayloadAction<string[]>) => {
            action.payload.forEach((userId: string) => {
                delete state.boardUsers[userId]
            })
        },
        followBlock: (state, action: PayloadAction<Subscription>) => {
            state.blockSubscriptions.push(action.payload)
        },
        unfollowBlock: (state, action: PayloadAction<Subscription>) => {
            const oldSubscriptions = state.blockSubscriptions
            state.blockSubscriptions = oldSubscriptions.filter((subscription) => subscription.blockId !== action.payload.blockId)
        },
        patchProps: (state, action: PayloadAction<Record<string, string>>) => {
            if (state.me) {
                state.me.props = parseUserProps(action.payload)
            }
        },
    },
    extraReducers: (builder) => {
        builder.addCase(fetchMe.fulfilled, (state, action) => {
            state.me = action.payload || null
            if (state.me) {
                state.me.props = parseUserProps(state.me.props)
            }
            state.loggedIn = Boolean(state.me)
        })
        builder.addCase(fetchMe.rejected, (state) => {
            state.me = null
            state.loggedIn = false
        })

        // TODO: change this when the initial load is complete
        // builder.addCase(initialLoad.fulfilled, (state, action) => {
        //     state.boardUsers = action.payload.boardUsers.reduce((acc: {[key: string]: IUser}, user: IUser) => {
        //         acc[user.id] = user
        //         return acc
        //     }, {})
        // })

        builder.addCase(fetchUserBlockSubscriptions.fulfilled, (state, action) => {
            state.blockSubscriptions = action.payload
        })
    },
})

export const {setMe, setBoardUsers, removeBoardUsersById, addBoardUsers, followBlock, unfollowBlock, patchProps} = usersSlice.actions
export const {reducer} = usersSlice

export const getMe = (state: RootState): IUser|null => state.users.me
export const getLoggedIn = (state: RootState): boolean|null => state.users.loggedIn
export const getBoardUsers = (state: RootState): {[key: string]: IUser} => state.users.boardUsers

export const getBoardUsersList = createSelector(
    getBoardUsers,
    (boardUsers) => Object.values(boardUsers).sort((a, b) => a.username.localeCompare(b.username)),
)

export const getUser = (userId: string): (state: RootState) => IUser|undefined => {
    return (state: RootState): IUser|undefined => {
        const users = getBoardUsers(state)
        return users[userId]
    }
}

export const getOnboardingTourStarted = createSelector(
    getMe,
    (me): boolean => {
        if (!me) {
            return false
        }

        return Boolean(me.props?.focalboard_onboardingTourStarted)
    },
)

export const getOnboardingTourStep = createSelector(
    getMe,
    (me): string => {
        if (!me) {
            return ''
        }

        return me.props?.focalboard_onboardingTourStep
    },
)

export const getOnboardingTourCategory = createSelector(
    getMe,
    (me): string => (me ? me.props?.focalboard_tourCategory : ''),
)

export const getCloudMessageCanceled = createSelector(
    getMe,
    (me): boolean => {
        if (!me) {
            return false
        }
        if (me.id === 'single-user') {
            return UserSettings.hideCloudMessage
        }
        return Boolean(me.props?.focalboard_cloudMessageCanceled)
    },
)

export const getVersionMessageCanceled = createSelector(
    getMe,
    (me): boolean => {
        if (versionProperty && me){
            if (me.id === 'single-user') {
                return true
            }
            return Boolean(me.props[versionProperty])    
        }
        return true
    },
)

export const getCardLimitSnoozeUntil = createSelector(
    getMe,
    (me): number => {
        if (!me) {
            return 0
        }
        try {
            return parseInt(me.props?.focalboard_cardLimitSnoozeUntil, 10) || 0
        } catch (_) {
            return 0
        }
    },
)

export const getCardHiddenWarningSnoozeUntil = createSelector(
    getMe,
    (me): number => {
        if (!me) {
            return 0
        }
        try {
            return parseInt(me.props?.focalboard_cardHiddenWarningSnoozeUntil, 10) || 0
        } catch (_) {
            return 0
        }
    },
)
