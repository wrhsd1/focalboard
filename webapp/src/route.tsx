// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react'
import {
    Redirect,
    Route,
} from 'react-router-dom'

import {Utils} from './utils'
import {getLoggedIn, getMe} from './store/users'
import {useAppSelector} from './store/hooks'
import {UserSettingKey} from './userSettings'
import {IUser, UserPropPrefix} from './user'
import {getClientConfig} from './store/clientConfig'
import {ClientConfig} from './config/clientConfig'

type RouteProps = {
    path: string|string[]
    exact?: boolean
    render?: (props: any) => React.ReactElement
    component?: React.ComponentType
    children?: React.ReactElement
    getOriginalPath?: (match: any) => string
    loginRequired?: boolean
}

function FBRoute(props: RouteProps) {
    const loggedIn = useAppSelector<boolean|null>(getLoggedIn)
    const me = useAppSelector<IUser|null>(getMe)
    const clientConfig = useAppSelector<ClientConfig>(getClientConfig)

    let redirect: React.ReactNode = null
    const disableTour = clientConfig?.featureFlags?.disableTour || false

    const showWelcomePage = !disableTour &&
        Utils.isFocalboardPlugin() &&
        (me?.id !== 'single-user') &&
        props.path !== '/welcome' &&
        loggedIn === true &&
        !me?.props[UserPropPrefix + UserSettingKey.WelcomePageViewed]

    if (showWelcomePage) {
        redirect = ({match}: any) => {
            if (props.getOriginalPath) {
                return <Redirect to={`/welcome?r=${props.getOriginalPath!(match)}`}/>
            }
            return <Redirect to='/welcome'/>
        }
    }

    if (redirect === null && loggedIn === false && props.loginRequired) {
        redirect = ({match}: any) => {
            if (props.getOriginalPath) {
                let redirectUrl = '/' + Utils.buildURL(props.getOriginalPath!(match))
                if (redirectUrl.indexOf('//') === 0) {
                    redirectUrl = redirectUrl.slice(1)
                }
                const loginUrl = `/error?id=not-logged-in&r=${encodeURIComponent(redirectUrl)}`
                return <Redirect to={loginUrl}/>
            }
            return <Redirect to='/error?id=not-logged-in'/>
        }
    }

    return (
        <Route
            path={props.path}
            render={props.render}
            component={props.component}
            exact={props.exact}
        >
            {redirect || props.children}
        </Route>
    )
}

export default React.memo(FBRoute)
