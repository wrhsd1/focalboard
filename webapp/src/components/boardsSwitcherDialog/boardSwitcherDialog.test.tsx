// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'

import {MockStoreEnhanced} from "redux-mock-store"
import {mocked} from 'jest-mock'

import {Provider as ReduxProvider} from 'react-redux'

import {render, screen, act} from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import {createMemoryHistory, History} from "history"

import {Router} from "react-router-dom"

import octoClient from '../../octoClient'

import {Team} from "../../store/teams"
import {TestBlockFactory} from "../../test/testBlockFactory"

import {mockStateStore, wrapDNDIntl} from "../../testUtils"

import {Board, createBoard} from '../../../../webapp/src/blocks/board'

import {Utils} from '../../../../webapp/src/utils'

import BoardSwitcherDialog from "./boardSwitcherDialog"

jest.mock('../../octoClient')
const mockedOctoClient = mocked(octoClient, true)
jest.mock('../../../../webapp/src/utils')
const mockedUtils = mocked(Utils, true)

const wait = (ms: number) => {
    return new Promise<void>((resolve) => {
        setTimeout(resolve, ms)
    })
}

describe('component/BoardSwitcherDialog', () => {
    const team1: Team = {
        id: 'team-id-1',
        title: 'Dunder Mifflin',
        signupToken: '',
        updateAt: 0,
        modifiedBy: 'michael-scott',
    }

    const team2: Team = {
        id: 'team-id-2',
        title: 'Michael Scott Paper Company',
        signupToken: '',
        updateAt: 0,
        modifiedBy: 'michael-scott',
    }

    const me = TestBlockFactory.createUser()

    const state = {
        users: {
            me: me,
        },
        teams: {
            allTeams: [team1, team2],
            current: team1,
        },
        sidebar: {
            categoryAttributes: [],
        }
    }

    let store:MockStoreEnhanced<unknown, unknown>
    let history: History


    beforeEach(() => {
        store = mockStateStore([], state)
        history = createMemoryHistory()
    })


    test('base case', () => {
        const onCloseHandler = jest.fn()
        const component = wrapDNDIntl(
            <Router history={history}>
                <ReduxProvider store={store}>
                    <BoardSwitcherDialog onClose={onCloseHandler}/>
                </ReduxProvider>
            </Router>
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })

    test('find one, default category', async () => {
        const onCloseHandler = jest.fn()
        const newBoard = createBoard({id: 'found-board', title: 'hello', teamId: 'team-id-1'} as Board)

        mockedOctoClient.searchAll.mockResolvedValue([newBoard])
        mockedUtils.uuid.mockReturnValue('Really a UUID')

        const component = wrapDNDIntl(
            <Router history={history}>
                <ReduxProvider store={store}>
                    <BoardSwitcherDialog onClose={onCloseHandler}/>
                </ReduxProvider>
            </Router>
        )

        const {container} = render(component)
        await act(async () => {
            const inputElement = screen.getByPlaceholderText('Search for boards')
            await userEvent.type(inputElement, 'test')
            await wait(300)
        })
        expect(container).toMatchSnapshot()
    })

    test('find one, with custom category', async () => {
        const onCloseHandler = jest.fn()

        const myState = {
            users: {
                me: me,
            },
            teams: {
                allTeams: [team1, team2],
                current: team1,
            },
            sidebar: {
                categoryAttributes: [{
                    name: 'TestCategory',
                    boardIDs: ['found-board']
                },],
            }
        }
    
        const myStore = mockStateStore([], myState)

        const newBoard = createBoard({id: 'found-board', title: 'hello', teamId: 'team-id-1'} as Board)

        mockedOctoClient.searchAll.mockResolvedValue([newBoard])
        mockedUtils.uuid.mockReturnValue('Really a UUID')

        const component = wrapDNDIntl(
            <Router history={history}>
                <ReduxProvider store={myStore}>
                    <BoardSwitcherDialog onClose={onCloseHandler}/>
                </ReduxProvider>
            </Router>
        )

        const {container} = render(component)
        await act(async () => {
            const inputElement = screen.getByPlaceholderText('Search for boards')
            await userEvent.type(inputElement, 'test')
            await wait(300)
        })
        expect(container).toMatchSnapshot()
    })
})
