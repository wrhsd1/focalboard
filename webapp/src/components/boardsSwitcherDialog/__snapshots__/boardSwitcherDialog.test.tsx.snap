// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/BoardSwitcherDialog base case 1`] = `
<div>
  <div
    class="Dialog dialog-back BoardSwitcherDialog"
  >
    <div
      class="backdrop"
    />
    <div
      class="wrapper"
    >
      <div
        class="dialog"
        role="dialog"
      >
        <div
          class="toolbar"
        >
          <button
            aria-label="Close dialog"
            title="Close dialog"
            type="button"
          >
            <i
              class="CompassIcon icon-close CloseIcon"
            />
          </button>
          <div
            class="toolbar--right"
          />
        </div>
        <div
          class="BoardSwitcherDialogBody"
        >
          <div
            class="head"
          >
            <h3
              class="text-heading4"
            >
              Find Boards
            </h3>
            <h5>
              Type to find a board. Use 
              <b>
                UP/DOWN
              </b>
               to browse. 
              <b>
                ENTER
              </b>
               to select, 
              <b>
                ESC
              </b>
               to dismiss
            </h5>
            <div
              class="queryWrapper"
            >
              <i
                class="CompassIcon icon-magnify MagnifyIcon"
              />
              <input
                class="searchQuery"
                maxlength="100"
                placeholder="Search for boards"
                type="text"
              />
            </div>
          </div>
          <div
            class="searchResults"
          >
            
            
            <div
              class="noResults introScreen"
            >
              <div
                class="iconWrapper"
              >
                <i
                  class="CompassIcon icon-magnify MagnifyIcon"
                />
              </div>
              <h4
                class="text-heading4"
              >
                Search for boards
              </h4>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`component/BoardSwitcherDialog find one, default category 1`] = `
<div>
  <div
    class="Dialog dialog-back BoardSwitcherDialog"
  >
    <div
      class="backdrop"
    />
    <div
      class="wrapper"
    >
      <div
        class="dialog"
        role="dialog"
      >
        <div
          class="toolbar"
        >
          <button
            aria-label="Close dialog"
            title="Close dialog"
            type="button"
          >
            <i
              class="CompassIcon icon-close CloseIcon"
            />
          </button>
          <div
            class="toolbar--right"
          />
        </div>
        <div
          class="BoardSwitcherDialogBody"
        >
          <div
            class="head"
          >
            <h3
              class="text-heading4"
            >
              Find Boards
            </h3>
            <h5>
              Type to find a board. Use 
              <b>
                UP/DOWN
              </b>
               to browse. 
              <b>
                ENTER
              </b>
               to select, 
              <b>
                ESC
              </b>
               to dismiss
            </h5>
            <div
              class="queryWrapper"
            >
              <i
                class="CompassIcon icon-magnify MagnifyIcon"
              />
              <input
                class="searchQuery"
                maxlength="100"
                placeholder="Search for boards"
                type="text"
              />
            </div>
          </div>
          <div
            class="searchResults"
          >
            <div
              class="searchResult"
              tabindex="-1"
            >
              <div
                class="blockSearchResult"
              >
                <i
                  class="CompassIcon icon-lock-outline LockOutlineIcon"
                />
                <div
                  class="resultTitle"
                >
                  <span>
                    hello
                  </span>
                </div>
                <div
                  class="categoryTitle"
                >
                  <span
                    class="ml-2 text-light"
                  >
                    Boards
                  </span>
                </div>
                <span
                  class="teamTitle"
                >
                  Dunder Mifflin
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`component/BoardSwitcherDialog find one, with custom category 1`] = `
<div>
  <div
    class="Dialog dialog-back BoardSwitcherDialog"
  >
    <div
      class="backdrop"
    />
    <div
      class="wrapper"
    >
      <div
        class="dialog"
        role="dialog"
      >
        <div
          class="toolbar"
        >
          <button
            aria-label="Close dialog"
            title="Close dialog"
            type="button"
          >
            <i
              class="CompassIcon icon-close CloseIcon"
            />
          </button>
          <div
            class="toolbar--right"
          />
        </div>
        <div
          class="BoardSwitcherDialogBody"
        >
          <div
            class="head"
          >
            <h3
              class="text-heading4"
            >
              Find Boards
            </h3>
            <h5>
              Type to find a board. Use 
              <b>
                UP/DOWN
              </b>
               to browse. 
              <b>
                ENTER
              </b>
               to select, 
              <b>
                ESC
              </b>
               to dismiss
            </h5>
            <div
              class="queryWrapper"
            >
              <i
                class="CompassIcon icon-magnify MagnifyIcon"
              />
              <input
                class="searchQuery"
                maxlength="100"
                placeholder="Search for boards"
                type="text"
              />
            </div>
          </div>
          <div
            class="searchResults"
          >
            <div
              class="searchResult"
              tabindex="-1"
            >
              <div
                class="blockSearchResult"
              >
                <i
                  class="CompassIcon icon-lock-outline LockOutlineIcon"
                />
                <div
                  class="resultTitle"
                >
                  <span>
                    hello
                  </span>
                </div>
                <div
                  class="categoryTitle"
                >
                  <span
                    class="ml-2 text-light"
                  >
                    TestCategory
                  </span>
                </div>
                <span
                  class="teamTitle"
                >
                  Dunder Mifflin
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
